### ISSUE TITLE
CPU consumption still increased

### ISSUE THREAD

**Issue by alxark** [grafana/agent#5060]
### What's wrong?

Updated to 0.36.0 version but still faces high CPU consumption on nodes without high load. 

```
grafana+ 1854009 31.5  2.9 1567504 57952 ?       Ssl  07:37  17:57 /usr/bin/grafana-agent-flow run --server.http.listen-addr=10.20.1.72:12345 --storage.path=/var/lib/grafana-agent-flow /etc/grafana-agent-flow.river
```

This host contains only 3 monitored containers. Less than 10 messages / second (during dump time it was almost 1 message / sec).

Metrics after several minutes of uptime:
```
# HELP zm_container_lines total number of log lines for container
# TYPE zm_container_lines counter
zm_container_lines{component_id="loki.process.preprocess"} 132
```

### Steps to reproduce

Start several docker containers with message generating.

### System information

Linux 5.15.0-73-generic #80-Ubuntu SMP Mon May 15 15:18:26 UTC 2023 x86_64 x86_64 x86_64 GNU/Linux

### Software version

v0.36.0

### Configuration

```text
discovery.docker "containers" {
    host = "unix:///var/run/docker.sock"
}

loki.source.docker "default" {
    host       = "unix:///var/run/docker.sock"

    targets    = discovery.relabel.logs.output

    forward_to = [ loki.process.preprocess.receiver ]
}

discovery.relabel "logs" {
    targets = discovery.docker.containers.targets

    rule {
        source_labels = ["__meta_docker_container_label_job" ]
        separator     = "/"
        target_label  = "job"
        action        = "replace"
    }
}

loki.process "preprocess" {
    forward_to = [ loki.write.loki.receiver ]

    stage.drop {
        older_than  = "10m"
        drop_counter_reason = "record is too old"
    }

    stage.metrics {
        metric.counter {
            name        = "container_lines"
            description = "total number of log lines for container"
            prefix      = "zm_"
            source      = "job"

            match_all         = true
            action            = "inc"
            max_idle_duration = "24h"
        }
    }
}

loki.write "loki" {
    endpoint {
        url = "http://grafana.htz:3100/loki/api/v1/push"
    }
}
```

**Comment by alxark** [comment]
CPU profiling added: 

[cpu.pprof.zip](https://github.com/grafana/agent/files/12497544/cpu.pprof.zip)

**Comment by rfratto** [comment]
Thanks for the profile. Looking at the profile, 33% of the profiled time is spent in `fmt.Sprintf`. From the code, that's clearly when generating debug logs, even if the log lines aren't written. 

Promtail doesn't have this problem, because they stopped logging those lines except in unit tests.

We have a few paths here:

* Bring over the same behavior while implementing #4836: only emit the log lines in unit tests.
* Change the log events to not use `fmt.Sprintf` and move information to fields instead so that there's no allocations. 
* Remove the log lines entirely, since they're only actually used in unit tests in Promtail. 

While in the provided provide this is only shown to be an issue in `stages.drop`, Promtail's code has this only-log-in-unit-tests check for all stages, so we'll want to do the same.

cc @tpaschalis

**Comment by alxark** [comment]
disabled stages.drop, but still CPU consumption above promtail levels.

[cpu.pprof.zip](https://github.com/grafana/agent/files/12555889/cpu.pprof.zip)

**Comment by alxark** [comment]
@rfratto  how to collect needed data?

**Comment by rfratto** [comment]
@alxark The /pprof/debug/heap endpoint will give you a memory profile.

**Comment by alxark** [comment]
Added. This profiler from different server. It consumes full core (100% of one core), and it's staging host (no hard work).

[profiler.mem.zip](https://github.com/grafana/agent/files/12564654/profiler.mem.zip)

**Comment by alxark** [comment]
this one is from staging server. CPU consumtion - 100% (full core). Only 2 containers, generating very small amount of logs + host metrics.

[profiler.mem.zip](https://github.com/grafana/agent/files/12564937/profiler.mem.zip)

**Comment by tpaschalis** [comment]
Looking at the heap profile `go tool pprof -http=:8081 profiler.mem`, it looks like there's something allocating _a lot_, but it isn't helpful in pinning down what
![image](https://github.com/grafana/agent/assets/17771679/e4d0bb87-435e-45b0-bcb8-ee1dac9df1fb)

I'm searching through what could lead to such a profile, but could you please grab and share another one?

**Comment by alxark** [comment]
Added more profiles. This host doesn't use drop.stage, but still CPU consumption around 27%. And this host doesn't generate too many messages.

[mem.profile.zip](https://github.com/grafana/agent/files/12647188/mem.profile.zip)
[cpu.profile.zip](https://github.com/grafana/agent/files/12647190/cpu.profile.zip)

```
discovery.docker "containers" {
    host = "unix:///var/run/docker.sock"
}

loki.source.docker "default" {
    host       = "unix:///var/run/docker.sock"

    targets    = discovery.relabel.logs.output

    forward_to = [ loki.process.preprocess.receiver ]
}

discovery.relabel "logs" {
    targets = discovery.docker.containers.targets

    rule {
        source_labels = ["__meta_docker_container_label_job" ]
        separator     = "/"
        target_label  = "job"
        action        = "replace"
    }
}

loki.process "preprocess" {
    forward_to = [ loki.write.loki.receiver ]

    stage.metrics {
        metric.counter {
            name        = "container_lines"
            description = "total number of log lines for container"
            prefix      = "zm_"
            source      = "job"

            match_all         = true
            action            = "inc"
            max_idle_duration = "24h"
        }
    }
}

loki.write "loki" {
    endpoint {
        url = "http://grafana.local:3100/loki/api/v1/push"
    }
}

```

**Comment by tpaschalis** [comment]
@alxark thanks for the update.

The memory profile still doesn't show symbols. How are you running the Agent? Is it just running the docker image from our Dockerhub, could it be that you're building from source and stripping the symbols, something else?

**Comment by alxark** [comment]
> @alxark thanks for the update.
> 
> The memory profile still doesn't show symbols. How are you running the Agent? Is it just running the docker image from our Dockerhub, could it be that you're building from source and stripping the symbols, something else?

we are using debian packages from github releases.

**Comment by tpaschalis** [comment]
cc @rfratto does it make sense for the debian packages to be stripped of their symbols?

**Comment by rfratto** [comment]
@tpaschalis Yeah, we [strip the symbols](https://github.com/grafana/agent/blob/6458bcdb296be3aaf3a14481adefc0939f220ff0/Makefile#L150) from release builds. This is historical (copied from Loki) but we should really consider reverting that decision. 

OTOH, it's not too much work for us, we just have to provide Go the binary and code to get symbols back:

```
$ go tool pprof -h                                                                                                      
usage:

Produce output in the specified format.

   pprof <format> [options] [binary] <source> ...
```

**Comment by sliemeobn** [comment]
tl;dr: it might be connected to large individual container log files.

I was seeing the same issue, and had a long overdue change planned to configure the default log driver (daemon.json) to roll files. (the docker default is "keep logs forever" as long as containers are running).

we ended up with large log files >3GB, since containers would run for a long time when no updates were planned. I also noticed that logs from these containers are often way behind (many minutes, sometimes hours).

on test machines (which are turned off and on much more frequently) the agent CPU-problem was a lot less visible.

so, I just had a maintenance window on our production env, changed the docker default to 25m/2 files, restarted all containers: agent CPU went from ~150% to ~2% and seems to stay there...

**Comment by tpaschalis** [comment]
@sliemeobn interesting context, thanks for adding this! @alxark could this be applicable to your case as well?

In any case, I was able to read the profile by providing the linux/amd64 v0.36.0 binary.

```
Showing top 10 nodes out of 177
      flat  flat%   sum%        cum   cum%
   11.39MB 18.76% 18.76%    11.39MB 18.76%  github.com/apache/arrow/go/v12/arrow/compute/internal/kernels.(*dictionaryHashState).Append
    5.70MB  9.38% 28.15%     5.70MB  9.38%  github.com/apache/arrow/go/v12/arrow/compute/internal/kernels.GetArithmeticFloatingPointKernels
    3.72MB  6.13% 34.28%     4.24MB  6.99%  github.com/scaleway/scaleway-sdk-go/scw.(*InvalidRequestError).ToInvalidArgumentsError
    2.50MB  4.12% 38.40%     2.50MB  4.12%  github.com/grafana/agent/component/prometheus/remotewrite.(*Component).Update
...
```

It looks like that a huge chunk the memory is taken up from packages initialization related to Apache Arrow. The only thing that pops out when I trace those dependencies is the mssql and snowflake integrations, which _don't_ look like they're part of your config, plus some metrics remote-writing which again don't look like they're part of the initial config.

So it's weird I can see nothing related to loki components or the docker driver taking up memory that has to be garbage collected and lead to CPU usage.

**Comment by thampiotr** [comment]
Related proposal to improve loggers performance specifically when given log level is disabled: https://github.com/grafana/agent/pull/5452

**Comment by thampiotr** [comment]
We've merged some performance improvements to Agent's logging, which can help address this: https://github.com/grafana/agent/issues/5060

I'll close the issue for now, but if you are able to test the fix and there are still problems, please let us know.

**Comment by ntimo** [comment]
Are there any news when this fix will be released? And can it be backported to 0.36?

**Comment by tpaschalis** [comment]
Hey @ntimo 👋  The proposed fix from @thampiotr will be released in v0.38 in November. We usually only backport documentation and security-related fixes, so I don't think we'd backport this until it surfaced a larger issue.

Furthermore, feel free to also check if you have the same issue reported above with large individual container log files and how it might impact things!
